### ISSUE TITLE
Bug handling AWS API Gateway events

### ISSUE THREAD

**Issue by ghost** [Ge0rg3/flask-parameter-validation#4]
Hi everyone!

This week I started getting errors in my api GET endpoints. After a lot of research I found out the reason.
I'm using AWS Lambda + AWS API Gateway to build some apis. I'm using the Flask + serverless_wsgi package to handle the API Gateway V2 requests and so having an common Flask api app, like code below:

```python

@app.route('/client/list')
@ValidateParameters()
def obter_clientes(page: int = Query(default=1),
                   per_page: int = Query(default=10)):
        return jsonify(data='Hello from lambda')


def main(event, context):
    return serverless_wsgi.handle_request(app, event, context)
```

and it work smoothly inside AWS Lambda functions. Normally API Gateway V2 sends an event to Lambda functions like this one:
```
{
  "version": "2.0",
  "routeKey": "$default",
  "rawPath": "/my/path",
  "rawQueryString": "parameter1=value1&parameter1=value2&parameter2=value",
  "headers": {
    "header1": "value1",
    "header2": "value1,value2"
  },
  "queryStringParameters": {
    "parameter1": "value1,value2",
    "parameter2": "value"
  },
  "requestContext": {
    "accountId": "123456789012",
    "apiId": "api-id",
    "domainName": "id.execute-api.us-east-1.amazonaws.com",
    "domainPrefix": "id",
    "http": {
      "method": "POST",
      "path": "/my/path",
      "protocol": "HTTP/1.1",
      "sourceIp": "IP",
      "userAgent": "agent"
    },
    "requestId": "id",
    "routeKey": "$default",
    "stage": "$default",
    "time": "12/Mar/2020:19:03:58 +0000",
    "timeEpoch": 1583348638390
  },
  "body": "Hello from Lambda",
  "pathParameters": {
    "parameter1": "value1"
  },
  "isBase64Encoded": false
}
```
I omitted some fields for brevity. This week for some reason the empty fields of the event is being removed by API Gateway before it  is send to the Lambda function. For example: In the GET Requests there is no body, so the body entry of the event json is omitted. In the past, all the fields were sent, but empty. The absence of the body entry in the event has generate an error in Flask Parameter Validation Package. Investigating the package code I noticed that the error is throw in the line `Json: request.json or {}` located inside the ValidateParameters class.

```python
class ValidateParameters:

    def __init__(self, error_handler=None):
        self.custom_error_handler = error_handler


    def __call__(self, f):
        """
        Parent flow for validating each required parameter
        """
        def nested_func(**kwargs):
            # Step 1 - Combine all flask input types to one dict
            request_inputs = {
                Route: kwargs.copy(),
                Json: request.json or {},
                Query: request.args.to_dict(),
                Form: request.form.to_dict(),
                File: request.files.to_dict()
            }
```

The `request.json` is throwing a 400 error because there is no body entry, even an empty body, in GET requests sent by API Gateway. Reading the flask documentation I found out there is this alternative `request.get_json(silent=True)` that does not throw an exception and return None instead. So I think replacing `Json: request.json or {}` for `request.get_json(silent=True) or {}` may resolve the issue.

I'll open a technical ticket in AWS support just to confirm this behavior in API Gateway events.

Sorry for the long text and thank you in advance!

**Comment by Ge0rg3** [comment]
Hey, thanks for the feedback on this! I've checked in the suggestion change to v2.0.5 to patch this for now. I'm thinking, though, that we should really be throwing an error in case of invalid json (and an empty string *is* invalid json, to be fair)?

**Comment by robson-araujo-solvimm** [comment]
It's not, neither `Content-Type` or `Content-Length` is being sent in GET requests

**Comment by Ge0rg3** [comment]
Hi, please can you double-check that? I've got the following file:
```python
from flask import Flask, jsonify
from typing import List, Optional
from flask_parameter_validation import ValidateParameters, Route, Json, Query

app = Flask(__name__)


@app.route('/client/list')
@ValidateParameters()
def obter_clientes(page: int = Query(default=1),
                   per_page: int = Query(default=10)):
        return jsonify(data='Hello from lambda')


def main(event, context):
    return serverless_wsgi.handle_request(app, event, context)
```
And I'm running it via `flask run`.
When running `curl http://127.0.0.1:5000/client/list`, I get the response `{"data":"Hello from lambda"}`.
However, when I run `curl http://127.0.0.1:5000/client/list -H 'Content-Type: application/json'`, I appropriately receive a 400 bad request.

**Comment by robson-araujo-solvimm** [comment]
Hi! I think the test must be made using the API Gateway event, the error is about the API Gateway event format. Here is the log of a GET request using API Gateway V2 + AWS Lambda with no  `Content-Type: application/json` or `body`:

```JSON
{
   "version":"2.0",
   "routeKey":"ANY /cliente/{proxy+}",
   "rawPath":"/cliente/list",
   "rawQueryString":"page=1&per_page=10",
   "headers":{
      "accept":"*/*",
      "accept-encoding":"gzip, deflate, br",
      "cache-control":"no-cache",
      "content-length":"0",
      "host":"bv7rljhfk5.execute-api.us-east-1.amazonaws.com",
      "postman-token":"bf91addd-7b60-4126-b2e2-aa7218bd5adb",
      "user-agent":"PostmanRuntime/7.28.4",
      "x-amzn-trace-id":"Root=1-6245ab92-79c7dc34571656a55d9944aa",
      "x-forwarded-for":"177.12.48.75",
      "x-forwarded-port":"443",
      "x-forwarded-proto":"https"
   },
   "queryStringParameters":{
      "page":"1",
      "per_page":"10"
   },
   "requestContext":{
      "accountId":"123642328289",
      "apiId":"bv7rljhfk5",
      "domainName":"bv7rljhfk5.execute-api.us-east-1.amazonaws.com",
      "domainPrefix":"bv7rljhfk5",
      "http":{
         "method":"GET",
         "path":"/cliente/list",
         "protocol":"HTTP/1.1",
         "sourceIp":"177.12.48.75",
         "userAgent":"PostmanRuntime/7.28.4"
      },
      "requestId":"P2e-7gNfoAMEMKw=",
      "routeKey":"ANY /cliente/{proxy+}",
      "stage":"$default",
      "time":"31/Mar/2022:13:24:34 +0000",
      "timeEpoch":1648733074596
   },
   "pathParameters":{
      "proxy":"list"
   },
   "isBase64Encoded":false
}
```
and the error is: `400 Bad Request: The browser (or proxy) sent a request that this server could not understand. `

**Comment by Ge0rg3** [comment]
This was fixed within #8 😄
